### PR TITLE
Added a warning window when clicking on 'cancel' on an expanded story with unsaved changes

### PR DESCRIPTION
**Pull request by GabrielRMuller**
image

### CODE REVIEW

File: app/assets/javascripts/components/story/ExpandedStory/ExpandedStoryControls.js [hunk 1]
```diff
@@ -1,33 +1,52 @@
...
+  constructor(props) {
+    super(props);
+
+    this.handleCancel = this.handleCancel.bind(this);
+  }
+
+  handleCancel() {
+    const { onCancel, isDirty } = this.props;
+
+    if (isDirty && !window.confirm(I18n.t('story unsaved changes'))) {
```
Comment (boniattirodrigo) [inline app/assets/javascripts/components/story/ExpandedStory/ExpandedStoryControls.js]: Don't you prefer to abstract the `!window.confirm(I18n.t('story unsaved changes'))` in a function or variable? It's a bit harder to figure out what it is doing.

File: app/assets/javascripts/components/story/ExpandedStory/ExpandedStoryControls.js [hunk 2]
```diff
@@ -1,33 +1,52 @@
...
-      <input className="cancel"
-        onClick={onCancel}
-        type="button"
-        value={I18n.t('cancel')}
-      />
-    </div>
-  );
+import React, { Component } from 'react';
+
+class ExpandedStoryControls extends Component {
```
Comment (gutofoletto) [inline app/assets/javascripts/components/story/ExpandedStory/ExpandedStoryControls.js]: Why did you change this functional component to a class?
It looks like you are not using internal state neither a lifecycle method.

File: app/assets/javascripts/components/story/ExpandedStory/ExpandedStoryControls.js [hunk 3]
```diff
@@ -1,33 +1,52 @@
...
+    this.handleCancel = this.handleCancel.bind(this);
+  }
+
+  handleCancel() {
+    const { onCancel, isDirty } = this.props;
+
+    if (isDirty && !window.confirm(I18n.t('story unsaved changes'))) {
+      return;
+    }
+    onCancel();
```
Comment (boniattirodrigo) [inline app/assets/javascripts/components/story/ExpandedStory/ExpandedStoryControls.js]: I think you could write just like that:

```
if (isDirty && hasNotUnsavedChanges) {
  onCancel();
}
```
Right?

File: app/assets/javascripts/components/story/ExpandedStory/ExpandedStoryControls.js [hunk 4]
```diff
@@ -1,33 +1,52 @@
...
+        <input className="save"
+          onClick={onSave}
+          type="button"
+          value={I18n.t('save')}
+          disabled={readOnly}
+        />
+
+        <input className="delete"
+          onClick={() => {
+            if (window.confirm(I18n.t('story destroy confirm'))){
```
Comment (boniattirodrigo) [inline app/assets/javascripts/components/story/ExpandedStory/ExpandedStoryControls.js]: It'll be clearer if you move this part to a function.

File: app/assets/javascripts/components/story/ExpandedStory/ExpandedStoryControls.js (hunk 3 above)
Comment (matheusazzi) [reply app/assets/javascripts/components/story/ExpandedStory/ExpandedStoryControls.js]: Without the context of it, I'd prefer`isDirty && savedAllChanges` since `hasNotUnsavedChanges` has "two negatives" and can be confusing.

File: spec/javascripts/components/story/expanded_story/expanded_story_controls_spec.js [hunk 5]
```diff
@@ -55,76 +55,125 @@ describe('<ExpandedStoryControls />', () => {
     });
   });
 
-  describe('when the user click on cancell', () => {
+  describe('when the user click on cancel', () => {
```
Comment (boniattirodrigo) [inline spec/javascripts/components/story/expanded_story/expanded_story_controls_spec.js]: Just a typo: when the user **clicks** on cancel

File: app/assets/javascripts/components/story/ExpandedStory/ExpandedStoryControls.js (hunk 2 above)
Comment (GabrielRMuller) [reply app/assets/javascripts/components/story/ExpandedStory/ExpandedStoryControls.js]: With the addition of more logic when canceling a story, I've created the method `handleCancel` and took the liberty of creating a class in order to encapsulate the logic into a single component.

File: app/assets/javascripts/components/story/ExpandedStory/ExpandedStoryControls.js (hunk 3 above)
Comment (boniattirodrigo) [reply app/assets/javascripts/components/story/ExpandedStory/ExpandedStoryControls.js]: Yeah, totally right. But I'm not sure if he's talking about `savedAllChanges`. Definitely, there's a better option.

Comment (GabrielRMuller) [reply app/assets/javascripts/components/story/ExpandedStory/ExpandedStoryControls.js]: Yeah, I can see how that can look a little bit confusing, I'm working on it, thanks!

File: spec/javascripts/components/story/expanded_story/expanded_story_controls_spec.js (hunk 5 above)
Comment (GabrielRMuller) [reply spec/javascripts/components/story/expanded_story/expanded_story_controls_spec.js]: good catch!

File: app/assets/javascripts/components/story/ExpandedStory/ExpandedStoryControls.js (hunk 2 above)
Comment (gutofoletto) [reply app/assets/javascripts/components/story/ExpandedStory/ExpandedStoryControls.js]: It was already a single component, you could just have used another function to `handleCancel`.
To put it simple, I just create class components when I need to use internal state or lifecycle methods, that's why I asked. But no problems there.

Comment (GabrielRMuller) [reply app/assets/javascripts/components/story/ExpandedStory/ExpandedStoryControls.js]: I see, thanks for the heads up!

Comment (talyssonoc) [reply app/assets/javascripts/components/story/ExpandedStory/ExpandedStoryControls.js]: @gutofoletto two things here:

1 - when reviewing, try to give context. For [your first comment ](https://github.com/Codeminer42/cm42-central/pull/508/files#r255657239) you make no parallel why internal state and lifecycle methods are related to functional components, you may commenting in code from people that didn't work with React before and PR reviews are meant to help people to learn.
2 - Avoid using "just" when suggesting changes, it may not be that simple for the person that is colaborating.

This being said, I disagree about creating two another functions (`handleDelete` and `handleCancel`) for that because functions give the impression they're pure when used for this kind of problem which it definitelly won't be because `window.confirm` isn't deterministic. Having a bunch of functions around won't make the code better than having a class that handles what we're doing in an encapsulated and easier to read way.

File: spec/javascripts/components/story/expanded_story/expanded_story_controls_spec.js [hunk 6]
```diff
@@ -55,76 +55,125 @@ describe('<ExpandedStoryControls />', () => {
...
+      expect(button.prop('disabled')).toBe(false);
+    });
+
+    it("don't disable delete button", () => {
+      const button = wrapper.find('.delete');
+
+      expect(button.prop('disabled')).toBe(false);
+    });
+  });
+});
```
Comment (talyssonoc) [inline spec/javascripts/components/story/expanded_story/expanded_story_controls_spec.js]: There's something weird with this indentation, it shouldn't have the same level as the one of the next line. Can you check it?

Comment (GabrielRMuller) [reply spec/javascripts/components/story/expanded_story/expanded_story_controls_spec.js]: will do.

File: spec/javascripts/components/story/expanded_story/expanded_story_controls_spec.js [hunk 7]
```diff
@@ -30,7 +30,7 @@ describe('<ExpandedStoryControls />', () => {
     });
   });
 
-  describe('when the user click on delete', () => {
+  describe('when the user clicks on delete', () => {
     beforeEach(function() {
```
Comment (talyssonoc) [inline spec/javascripts/components/story/expanded_story/expanded_story_controls_spec.js]: You can use an arrow function here instead of a regular function
```suggestion
     beforeEach(() => {
```

File: spec/javascripts/components/story/expanded_story/expanded_story_controls_spec.js [hunk 8]
```diff
@@ -55,76 +55,125 @@ describe('<ExpandedStoryControls />', () => {
     });
   });
 
-  describe('when the user click on cancell', () => {
+  describe('when the user clicks on cancel', () => {
+    beforeEach(function() {
```
Comment (talyssonoc) [inline spec/javascripts/components/story/expanded_story/expanded_story_controls_spec.js]:
```suggestion
    beforeEach(() => {
```

File: spec/javascripts/components/story/expanded_story/expanded_story_controls_spec.js [hunk 9]
```diff
@@ -55,76 +55,125 @@ describe('<ExpandedStoryControls />', () => {
...
+      sinon.stub(window, 'confirm')
+    });
+
+    afterEach(function() {
```
Comment (talyssonoc) [inline spec/javascripts/components/story/expanded_story/expanded_story_controls_spec.js]:
```suggestion
    afterEach(() => {
```

File: spec/javascripts/components/story/expanded_story/expanded_story_controls_spec.js [hunk 10]
```diff
@@ -55,76 +55,125 @@ describe('<ExpandedStoryControls />', () => {
...
+        />
+      );
 
-      it('disable save button', () => {
-        const button = wrapper.find('.save');
+      wrapper.find('.cancel').simulate('click');
 
-        expect(button.prop('disabled')).toBe(true);
-      });
+      expect(window.confirm).toHaveBeenCalled();
```
Comment (talyssonoc) [inline spec/javascripts/components/story/expanded_story/expanded_story_controls_spec.js]: Can we add an assertion to guarantee `handleCancel` isn't being called as well?

File: spec/javascripts/components/story/expanded_story/expanded_story_controls_spec.js [hunk 11]
```diff
@@ -55,76 +55,125 @@ describe('<ExpandedStoryControls />', () => {
...
           readOnly={false}
-          onCancel={onCancel}
+          onCancel={handleCancel}
         />
       );
 
       wrapper.find('.cancel').simulate('click');
 
-      expect(onCancel).toHaveBeenCalled();
+      expect(handleCancel).toHaveBeenCalled();
```
Comment (talyssonoc) [inline spec/javascripts/components/story/expanded_story/expanded_story_controls_spec.js]: We should add an assertion that checks `window.confirm` is not being called as well, what do you think?

File: spec/javascripts/components/story/expanded_story/expanded_story_controls_spec.js (hunk 6 above)
Comment (GabrielRMuller) [reply spec/javascripts/components/story/expanded_story/expanded_story_controls_spec.js]: fixed!

File: spec/javascripts/components/story/expanded_story/expanded_story_controls_spec.js [hunk 12]
```diff
@@ -55,20 +55,69 @@ describe('<ExpandedStoryControls />', () => {
...
+        />
+      );
+
+      wrapper.find('.cancel').simulate('click');
+
+      expect(window.confirm).not.toHaveBeenCalled();
+    });
+
+    describe('when the user clicks cancel on window.confirm', () => {
+      it('does not call onCancel', () => {
```
Comment (talyssonoc) [inline spec/javascripts/components/story/expanded_story/expanded_story_controls_spec.js]: This test doesn't do much anymore, right?
